### ISSUE TITLE
Many warnings with -Wconversion

### ISSUE THREAD

**Issue by nvibd** [NVIDIA/thrust#1478]
We're seeing a lot of warnings on Linux, host compiler GCC 9.3.0 with -Wconversion.

For a basic example, compile:
```
#include <thrust/device_vector.h>
int main() {
  thrust::device_vector<int> a;
  return 0;
}
```

With: `nvcc main.cu -Xcompiler=-Wconversion`.

This will result in around 1000 lines of output warnings.
Tested with the Thrust versions in CUDA 11.0, 11.3, and the latest commit on main on this repository (`776bbc4d`).

**Comment by allisonvacanti** [comment]
We should add this flag to Thrust and CUB's CI build system for all targets.
